File: agent-adapter/ts/src/backends/openhands.ts
// =============================================================================
// TurboFlow Agent Adapter — OpenHands Backend
//
// Wraps OpenHands (https://github.com/All-Hands-AI/OpenHands).
// MIT licensed, multi-model, MCP-compatible.
// Runs via CLI or Docker container.
// =============================================================================

import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { homedir } from 'node:os';
import { AgentBackend } from '../backend.js';
import { run, commandExists } from '../process.js';
import type {
  BackendCapabilities,
  ExecOptions,
  ExecResult,
  HealthCheckResult,
  McpServer,
} from '../types.js';

const DEFAULT_IMAGE = 'ghcr.io/all-hands-ai/openhands:latest';

export class OpenHandsBackend extends AgentBackend {
  readonly name = 'OpenHands';
  readonly description = 'Open-source AI software engineer (MIT license)';
  readonly url = 'https://github.com/All-Hands-AI/OpenHands';
  readonly license = 'MIT';

  readonly capabilities: BackendCapabilities = {
    mcp: true,
    agentTeams: false,
    multiModel: true,
    interactive: false,
    headless: true,
    streaming: true,
    bedrock: false,
    worktrees: true,
  };

  private readonly image: string;

  constructor() {
    super('openhands');
    this.image = process.env.OPENHANDS_IMAGE ?? DEFAULT_IMAGE;
  }

  resolveModel(model: string): string {
    const tierMap: Record<string, string> = {
      opus: 'anthropic/claude-opus-4-20250514',
      sonnet: 'anthropic/claude-sonnet-4-20250514',
      haiku: 'anthropic/claude-haiku-4-5-20251001',
    };
    return tierMap[model] ?? model;
  }

  private detectApiKey(model: string): string | undefined {
    if (process.env.LLM_API_KEY) return process.env.LLM_API_KEY;
    if (model.startsWith('anthropic/')) return process.env.ANTHROPIC_API_KEY;
    if (model.startsWith('openai/')) return process.env.OPENAI_API_KEY;
    return process.env.ANTHROPIC_API_KEY ?? process.env.OPENAI_API_KEY;
  }

  async exec(prompt: string, options: ExecOptions = {}): Promise<ExecResult> {
    const model = options.model
      ? this.resolveModel(options.model)
      : process.env.LLM_MODEL ?? 'anthropic/claude-sonnet-4-20250514';

    const apiKey = this.detectApiKey(model);
    if (!apiKey) {
      return {
        exitCode: 1,
        stdout: '',
        stderr: 'No API key found. Set LLM_API_KEY, ANTHROPIC_API_KEY, or OPENAI_API_KEY.',
        durationMs: 0,
        timedOut: false,
      };
    }

    // Add file context to prompt
    let fullPrompt = prompt;
    if (options.files?.length) {
      fullPrompt = `Focus on these files: ${options.files.join(', ')}. ${prompt}`;
    }

    const env = {
      LLM_MODEL: model,
      LLM_API_KEY: apiKey,
      ...options.env,
    };

    // Prefer CLI
    if (await commandExists('openhands')) {
      return run({
        command: 'openhands',
        args: ['run', '--task', fullPrompt, ...(options.extraArgs ?? [])],
        cwd: options.cwd,
        env,
        timeout: options.timeout,
        stream: true,
      });
    }

    // Fallback to Docker
    if (await commandExists('docker')) {
      const cwd = options.cwd ?? process.cwd();
      return run({
        command: 'docker',
        args: [
          'run', '--rm',
          '-e', `LLM_MODEL=${model}`,
          '-e', `LLM_API_KEY=${apiKey}`,
          '-v', `${cwd}:/workspace:rw`,
          this.image,
          'python', '-m', 'openhands.core.main',
          '-t', fullPrompt,
          ...(options.extraArgs ?? []),
        ],
        timeout: options.timeout,
        stream: true,
      });
    }

    return {
      exitCode: 1,
      stdout: '',
      stderr: 'Neither openhands CLI nor Docker found. Run: pip install openhands',
      durationMs: 0,
      timedOut: false,
    };
  }

  async version(): Promise<string | null> {
    if (await commandExists('openhands')) {
      const result = await run({ command: 'openhands', args: ['--version'], timeout: 10_000 });
      return result.exitCode === 0 ? result.stdout.trim() : null;
    }

    if (await commandExists('docker')) {
      const result = await run({
        command: 'docker',
        args: ['image', 'inspect', this.image, '--format', '{{index .RepoTags 0}}'],
        timeout: 10_000,
      });
      return result.exitCode === 0 ? `docker: ${result.stdout.trim()}` : null;
    }

    return null;
  }

  async isInstalled(): Promise<boolean> {
    if (await commandExists('openhands')) return true;

    // Check if Docker image is pulled
    if (await commandExists('docker')) {
      const result = await run({
        command: 'docker',
        args: ['image', 'inspect', this.image],
        timeout: 10_000,
      });
      return result.exitCode === 0;
    }

    return false;
  }

  async healthCheck(): Promise<HealthCheckResult> {
    const installed = await this.isInstalled();
    const version = installed ? await this.version() : null;
    const warnings: string[] = [];
    const details: Record<string, string | boolean> = {};

    details.cliAvailable = await commandExists('openhands');
    details.dockerAvailable = await commandExists('docker');

    if (details.dockerAvailable) {
      const imgResult = await run({
        command: 'docker',
        args: ['image', 'inspect', this.image],
        timeout: 10_000,
      });
      details.imagePulled = imgResult.exitCode === 0;
      if (!details.imagePulled) {
        warnings.push(`Docker image not pulled. Run: docker pull ${this.image}`);
      }
    }

    // Check providers
    const providers: string[] = [];
    if (process.env.ANTHROPIC_API_KEY) providers.push('Anthropic');
    if (process.env.OPENAI_API_KEY) providers.push('OpenAI');
    if (process.env.LLM_MODEL) details.model = process.env.LLM_MODEL;

    if (providers.length === 0 && !process.env.LLM_API_KEY) {
      warnings.push('No API key set. Set LLM_API_KEY, ANTHROPIC_API_KEY, or OPENAI_API_KEY');
    }

    details.providers = providers.join(', ') || 'none';

    return {
      installed,
      version,
      provider: providers[0] ?? null,
      details,
      warnings,
    };
  }

  async install(): Promise<void> {
    this.log.info('Installing OpenHands...');

    // Try pip
    if (await commandExists('pip') || await commandExists('pip3')) {
      const pipCmd = await commandExists('pip3') ? 'pip3' : 'pip';
      const result = await run({
        command: pipCmd,
        args: ['install', '--user', 'openhands'],
        timeout: 180_000,
        stream: true,
      });
      if (result.exitCode === 0) {
        this.log.info('OpenHands installed (pip)');
        return;
      }
    }

    // Try Docker pull
    if (await commandExists('docker')) {
      const result = await run({
        command: 'docker',
        args: ['pull', this.image],
        timeout: 300_000,
        stream: true,
      });
      if (result.exitCode === 0) {
        this.log.info('OpenHands Docker image pulled');
        return;
      }
    }

    throw new Error(`OpenHands installation failed. Try: pip install openhands or docker pull ${this.image}`);
  }

  // ── MCP support via config file ─────────────────────────────────────────

  private get mcpConfigPath(): string {
    return join(homedir(), '.openhands', 'mcp_servers.json');
  }

  private readMcpConfig(): Record<string, { command: string; args?: string[] }> {
    try {
      if (existsSync(this.mcpConfigPath)) {
        return JSON.parse(readFileSync(this.mcpConfigPath, 'utf-8'));
      }
    } catch {
      this.log.warn('Failed to read MCP config');
    }
    return {};
  }

  private writeMcpConfig(config: Record<string, { command: string; args?: string[] }>): void {
    const dir = join(homedir(), '.openhands');
    mkdirSync(dir, { recursive: true });
    writeFileSync(this.mcpConfigPath, JSON.stringify(config, null, 2));
  }

  async mcpAdd(server: McpServer): Promise<void> {
    const config = this.readMcpConfig();
    config[server.name] = { command: server.command, args: server.args };
    this.writeMcpConfig(config);
    this.log.info(`MCP server '${server.name}' registered (openhands config)`);
  }

  async mcpRemove(name: string): Promise<void> {
    const config = this.readMcpConfig();
    delete config[name];
    this.writeMcpConfig(config);
  }

  async mcpList(): Promise<McpServer[]> {
    const config = this.readMcpConfig();
    return Object.entries(config).map(([name, val]) => ({
      name,
      command: val.command,
      args: val.args,
    }));
  }
}

File: agent-adapter/python/turboflow_adapter/backends/openhands.py
"""OpenHands backend — wraps the OpenHands CLI or Docker container."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from turboflow_adapter.backend import AgentBackend
from turboflow_adapter.process import run, command_exists
from turboflow_adapter.types import (
    BackendCapabilities,
    ExecOptions,
    ExecResult,
    HealthCheckResult,
    McpServer,
)

DEFAULT_IMAGE = "ghcr.io/all-hands-ai/openhands:latest"


class OpenHandsBackend(AgentBackend):
    def __init__(self) -> None:
        super().__init__("openhands")
        self._image = os.environ.get("OPENHANDS_IMAGE", DEFAULT_IMAGE)

    @property
    def name(self) -> str:
        return "OpenHands"

    @property
    def description(self) -> str:
        return "Open-source AI software engineer (MIT license)"

    @property
    def url(self) -> str:
        return "https://github.com/All-Hands-AI/OpenHands"

    @property
    def license(self) -> str:
        return "MIT"

    @property
    def capabilities(self) -> BackendCapabilities:
        return BackendCapabilities(
            mcp=True, multi_model=True, headless=True, streaming=True, worktrees=True
        )

    def resolve_model(self, model: str) -> str:
        tier_map = {
            "opus": "anthropic/claude-opus-4-20250514",
            "sonnet": "anthropic/claude-sonnet-4-20250514",
            "haiku": "anthropic/claude-haiku-4-5-20251001",
        }
        return tier_map.get(model, model)

    def _detect_api_key(self, model: str) -> str | None:
        if key := os.environ.get("LLM_API_KEY"):
            return key
        if model.startswith("anthropic/"):
            return os.environ.get("ANTHROPIC_API_KEY")
        if model.startswith("openai/"):
            return os.environ.get("OPENAI_API_KEY")
        return os.environ.get("ANTHROPIC_API_KEY") or os.environ.get("OPENAI_API_KEY")

    async def exec(self, prompt: str, options: ExecOptions | None = None) -> ExecResult:
        opts = options or ExecOptions()
        model = (
            self.resolve_model(opts.model)
            if opts.model
            else os.environ.get("LLM_MODEL", "anthropic/claude-sonnet-4-20250514")
        )
        api_key = self._detect_api_key(model)
        if not api_key:
            return ExecResult(exit_code=1, stdout="", stderr="No API key found", duration_ms=0)

        full_prompt = prompt
        if opts.files:
            full_prompt = f"Focus on these files: {', '.join(opts.files)}. {prompt}"

        env = {"LLM_MODEL": model, "LLM_API_KEY": api_key, **(opts.env or {})}

        if command_exists("openhands"):
            return run(
                "openhands",
                args=["run", "--task", full_prompt],
                cwd=opts.cwd,
                env=env,
                timeout=opts.timeout,
                stream=True,
            )

        if command_exists("docker"):
            cwd = opts.cwd or os.getcwd()
            return run(
                "docker",
                args=[
                    "run",
                    "--rm",
                    "-e",
                    f"LLM_MODEL={model}",
                    "-e",
                    f"LLM_API_KEY={api_key}",
                    "-v",
                    f"{cwd}:/workspace:rw",
                    self._image,
                    "python",
                    "-m",
                    "openhands.core.main",
                    "-t",
                    full_prompt,
                ],
                timeout=opts.timeout,
                stream=True,
            )

        return ExecResult(
            exit_code=1, stdout="", stderr="Neither openhands CLI nor Docker found", duration_ms=0
        )

    async def version(self) -> str | None:
        if command_exists("openhands"):
            result = run("openhands", args=["--version"], timeout=10)
            return result.stdout.strip() if result.exit_code == 0 else None
        return None

    async def is_installed(self) -> bool:
        return command_exists("openhands") or command_exists("docker")

    async def health_check(self) -> HealthCheckResult:
        installed = await self.is_installed()
        ver = await self.version() if installed else None
        warnings: list[str] = []
        details: dict[str, str | bool] = {
            "cli": command_exists("openhands"),
            "docker": command_exists("docker"),
        }
        if (
            not os.environ.get("ANTHROPIC_API_KEY")
            and not os.environ.get("OPENAI_API_KEY")
            and not os.environ.get("LLM_API_KEY")
        ):
            warnings.append("No API key set")
        return HealthCheckResult(
            installed=installed, version=ver, provider=None, details=details, warnings=warnings
        )

    async def install(self) -> None:
        self.log.info("Installing OpenHands...")
        pip = "pip3" if command_exists("pip3") else "pip"
        result = run(pip, args=["install", "--user", "openhands"], timeout=180, stream=True)
        if result.exit_code == 0:
            return
        if command_exists("docker"):
            result = run("docker", args=["pull", self._image], timeout=300, stream=True)
            if result.exit_code == 0:
                return
        raise RuntimeError("OpenHands installation failed")

    # ── MCP via config file ──────────────────────────────────────────────

    @property
    def _mcp_config_path(self) -> Path:
        return Path.home() / ".openhands" / "mcp_servers.json"

    def _read_mcp_config(self) -> dict[str, Any]:
        try:
            if self._mcp_config_path.exists():
                result: dict[str, Any] = json.loads(self._mcp_config_path.read_text())
                return result
        except Exception:
            pass
        return {}

    def _write_mcp_config(self, config: dict[str, Any]) -> None:
        self._mcp_config_path.parent.mkdir(parents=True, exist_ok=True)
        self._mcp_config_path.write_text(json.dumps(config, indent=2))

    async def mcp_add(self, server: McpServer) -> None:
        config = self._read_mcp_config()
        config[server.name] = {"command": server.command, "args": server.args}
        self._write_mcp_config(config)
        self.log.info("MCP server '%s' registered (openhands)", server.name)

    async def mcp_remove(self, name: str) -> None:
        config = self._read_mcp_config()
        config.pop(name, None)
        self._write_mcp_config(config)
